### DISCUSSION TITLE
How to set up a discrete port which is popular for antenna applications in EM industrial software

### DISCUSSION THREAD

**Discussion by tmteryz** [NanoComp/meep#2457] [Q&A]
I need to calculate the return loss or S11 for antenna applications,  could anyone tell me how to set up the discrete port in meep ?
the discrete port is modeled by a lumped element, consisting of a current source with an inner impedance placed in parallel (see picture below) that excites and absorbs power.

![ae542987-393c-4b3e-93a5-80d6baa0efee](https://user-images.githubusercontent.com/124122684/229803790-f99a983f-8193-4baf-8973-26c8d642a2aa.png)

**Comment by smartalecH** [comment]
You could perform a [mode expansion](https://meep.readthedocs.io/en/latest/Python_Tutorials/Mode_Decomposition/) at your ports and compute the backward-traveling field coefficient (and from that compute S11 etc.)

The challenge with this is that it uses MPB to perform the mode decomposition, which does not work with metals. And RF transmission lines are typically metal...

It's been proposed for some time to generalize the feature for [arbitrary mode solvers](https://github.com/NanoComp/meep/issues/1319) (i.e. some that work with metals) but no one has pursued that path thoroughly yet.

**Reply by tmteryz** [reply]
Thank you for your reply. 
I have tried to use MPB to solve the mode under the metal/PEC boundary, but failed as stated in the MPB documentation.
So far, I haven't seen any concepts related to lumped-element in meep, so I'm not quite sure if it's feasible to develop discrete port functions for approximating electrically small excitation sources in the meep framework.

**Reply by smartalecH** [reply]
> I'm not quite sure if it's feasible to develop discrete port functions for approximating electrically small excitation sources in the meep framework.

The issue isn’t with meep. Meep handles metals and arbitrary boundaries just fine. And it can compute (with some post-processing) lumped element models too. 

The issue is the mode solver that meep currently relies on. Until that’s generalized, you probably won’t (easily) be able to do what you want.

**Reply by tmteryz** [reply]
i agree with you，but we're talking about the discrete port.......
the waveguide port requires a 2D eigenmode solver (like MPB),  while the discrete port is a current source with an inner impedance (e.g. 50 ohms).
